Fix return type inference for aggregate calls: consider number of items in GROUP BY clause

vlsi opened the pull request:
This fixes https://github.com/julianhyde/optiq/issues/231 (tested as `JdbcTest.testGroupByMax1IsNull`)

Side fixes:
- alter min/max return type inference to ARG0_NULLABLE_IF_EMPTY (nullable when no keys in group by)
- PushAggregateThroughUnionRule
- ability to log final plan in Prepare
- make emp.mgr nullable in `MockCatalogReader` (just to increase test coverage)

julianhyde commented on core/src/main/java/org/eigenbase/sql/type/SqlTypeUtil.java:
```diff
@@ -1103,6 +1103,29 @@ public static int findField(RelDataType type, String fieldName) {
   }
 
   /**
+   * Projects data types ou
```
fix comment, add param and return descriptions

julianhyde commented on core/src/test/java/net/hydromatic/optiq/tools/PlannerTest.java:
```diff
@@ -507,7 +507,7 @@ public MockJdbcTableScan(RelOptCluster cluster, RelOptTable table,
   /**
    * Test to determine whether de-correlation correctly removes CorrelatorRel.
    */
-  @Test public void testOldJoinStyleDeCorrelation() throws Exception {
```
please revert typo

vlsi committed:
Fix return type inference for aggregate calls: it should consider number of items in GROUP BY clause.
Side fixes:
  - alter min/max return type inference to ARG0_NULLABLE_IF_EMPTY (nullable when no keys in group by)
  - PushAggregateThroughUnionRule
  - ability to log final plan in Prepare
  - make emp.mgr nullable (just to increase test coverage)

vlsi committed:
Rework PullConstantsThroughAggregatesRule in terms of ProjectRel instead of CalcRel

julianhyde commented:
Merged as https://github.com/julianhyde/optiq/commit/61f2fdfcfb36f8052c9a3d645197cb649fbdf4e2.
